File: mobile/lib/data/services/wheel_mode.dart
import 'package:shared_preferences/shared_preferences.dart';

import 'controller_preset.dart';

/// How steering input is captured on the phone.
enum WheelMode {
  /// Finger-drag circular wheel; N degrees of finger rotation = full lock.
  rotatable('rotatable', 'Rotatable'),

  /// Phone tilt via gyroscope.
  gyro('gyro', 'Gyro');

  const WheelMode(this.wireValue, this.label);
  final String wireValue;
  final String label;

  static const String prefsKey = 'wheeldeck.wheel_mode';
  static const WheelMode fallback = rotatable;

  static WheelMode fromWireValue(String? v) => WheelMode.values
      .firstWhere((m) => m.wireValue == v, orElse: () => fallback);

  static Future<WheelMode> load() async {
    final prefs = await SharedPreferences.getInstance();
    return fromWireValue(prefs.getString(prefsKey));
  }

  Future<void> save() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(prefsKey, wireValue);
  }
}

/// Finger-rotation range mapping to full lock-to-lock, stored per game preset.
/// Gyro steering ignores degrees.
class RotationDegree {
  static const List<int> allowed = [180, 270, 900, 1080, 1800, 2520];
  static const int fallback = 900;

  static const String keyPrefix = 'wheeldeck.rotation_degree.';

  static String _key(GamePreset preset) => '$keyPrefix${preset.wireValue}';

  static Future<int> load(GamePreset preset) async {
    final prefs = await SharedPreferences.getInstance();
    final stored = prefs.getInt(_key(preset));
    if (stored == null || !allowed.contains(stored)) return fallback;
    return stored;
  }

  static Future<void> save(GamePreset preset, int degree) async {
    assert(allowed.contains(degree), 'Unsupported rotation degree: $degree');
    final prefs = await SharedPreferences.getInstance();
    await prefs.setInt(_key(preset), degree);
  }
}

File: android/app/src/main/kotlin/dev/fazrigading/wheeldeck/data/services/WheelMode.kt
package dev.fazrigading.wheeldeck.data.services

/// How steering input is captured on the phone.
enum class WheelMode(val wireValue: String, val label: String) {
    /** Finger-drag circular wheel; N degrees of finger rotation = full lock. */
    Rotatable("rotatable", "Rotatable"),

    /** Phone tilt via gyroscope. */
    Gyro("gyro", "Gyro");

    companion object {
        const val KEY = "wheeldeck.wheel_mode"

        /// Fresh installs drive the rotatable wheel.
        val fallback = Rotatable

        fun fromWireValue(value: String?): WheelMode =
            entries.firstOrNull { it.wireValue == value } ?: fallback
    }
}

/// Finger-rotation range mapping to full lock-to-lock, stored per game preset.
/// Gyro steering ignores degrees.
object RotationDegree {
    val allowed = listOf(180, 270, 900, 1080, 1800, 2520)
    const val FALLBACK = 900
    const val KEY_PREFIX = "wheeldeck.rotation_degree."

    fun key(preset: GamePreset) = "$KEY_PREFIX${preset.wireValue}"

    /// Stored value, or [FALLBACK] when unset or not one of [allowed].
    fun resolve(stored: Int?): Int = if (stored in allowed) stored!! else FALLBACK
}
